### PR TITLE
Webgraph adapter

### PR DESCRIPTION
**Pull request by d-michail**
This is a pull-request containing work done by @vigna on adapter classes for [WebGraph](http://webgraph.di.unimi.it/).
Together with our recent changes on the Graph interface, it allows several JGraphT algorithms (with possibly some modifications) to be executed on very large graphs.

----
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [x] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [x] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [x] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [x] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [x] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

### CODE REVIEW

**Comment by vigna** [comment]
I missed the part about checkstyle/javadoc compliance. Working on it...

**Comment by vigna** [comment]
BTW, I noticed that (following the JDK style) you don't put end-of-phrase periods in Javadoc's parameter/return value descriptions. We are used to put them in, but if you find it against your standards I can go through the code and eliminate them.

**Comment by d-michail** [comment]
I usually try to not include periods, but we never enforce this strictly. Thus, you will find both types in our codebase, sometimes even in the same class.

**Comment by jsichi** [comment]
This one can't be merged until the webgraph dependency is bumped up to a version with a compatible license.

**Comment by jsichi** [comment]
Sorry, my bad, looks like the Apache licensing is available as of 3.6.6.  So ignore my previous comment.

**Comment by jsichi** [comment]
Below is a list of the jars produced by `mvn package` off of the webgraph branch.  We need exclusions for the jung stuff at a minimum.

```
jvs@havnor:~/open/d-michail/jgrapht/jgrapht-dist/target/jgrapht-1.5.1-SNAPSHOT/lib$ ls -l
total 31828
-rw-rw-r-- 1 jvs jvs   337869 Oct 27 12:07 antlr4-runtime-4.8-1.jar
-rw-rw-r-- 1 jvs jvs   563060 Nov 28 14:25 commons-collections-20040616.jar
-rw-rw-r-- 1 jvs jvs   362679 Nov 28 14:25 commons-configuration-1.10.jar
-rw-rw-r-- 1 jvs jvs   285424 Nov 28 14:25 commons-io-2.8.0.jar
-rw-rw-r-- 1 jvs jvs   284220 Oct 27 12:08 commons-lang-2.6.jar
-rw-rw-r-- 1 jvs jvs   503880 Oct 27 12:07 commons-lang3-3.9.jar
-rw-rw-r-- 1 jvs jvs    60686 Nov 28 14:25 commons-logging-1.1.1.jar
-rw-rw-r-- 1 jvs jvs  2213560 Nov 28 14:25 commons-math3-3.6.1.jar
-rw-rw-r-- 1 jvs jvs   207030 Oct 27 12:07 commons-text-1.8.jar
-rw-rw-r-- 1 jvs jvs   500979 Nov 28 14:25 dsiutils-2.6.14.jar
-rw-rw-r-- 1 jvs jvs 19410584 Nov 28 14:25 fastutil-8.4.3.jar
-rw-rw-r-- 1 jvs jvs  2792264 Oct 27 11:39 guava-29.0-jre.jar
-rw-rw-r-- 1 jvs jvs  1159414 Nov 28 14:24 jgrapht-core-1.5.1-SNAPSHOT.jar
-rw-rw-r-- 1 jvs jvs    36078 Nov 28 14:25 jgrapht-demo-1.5.1-SNAPSHOT.jar
-rw-rw-r-- 1 jvs jvs     5711 Nov 28 14:25 jgrapht-ext-1.5.1-SNAPSHOT.jar
-rw-rw-r-- 1 jvs jvs    46327 Nov 28 14:25 jgrapht-guava-1.5.1-SNAPSHOT.jar
-rw-rw-r-- 1 jvs jvs   356675 Nov 28 14:24 jgrapht-io-1.5.1-SNAPSHOT.jar
-rw-rw-r-- 1 jvs jvs    73784 Nov 28 14:25 jgrapht-opt-1.5.1-SNAPSHOT.jar
-rw-rw-r-- 1 jvs jvs   832333 Oct 27 12:08 jgraphx-4.1.0.jar
-rw-rw-r-- 1 jvs jvs   152480 Oct 27 11:38 jheaps-0.13.jar
-rw-rw-r-- 1 jvs jvs    69344 Nov 28 14:25 jsap-2.1.jar
-rw-rw-r-- 1 jvs jvs   230204 Nov 28 14:25 jung-algorithms-2.1.1.jar
-rw-rw-r-- 1 jvs jvs    42221 Nov 28 14:25 jung-api-2.1.1.jar
-rw-rw-r-- 1 jvs jvs    79779 Nov 28 14:25 jung-io-2.1.1.jar
-rw-rw-r-- 1 jvs jvs   290339 Nov 28 14:25 logback-classic-1.2.3.jar
-rw-rw-r-- 1 jvs jvs   471901 Nov 28 14:25 logback-core-1.2.3.jar
-rw-rw-r-- 1 jvs jvs    41472 Nov 28 14:25 slf4j-api-1.7.30.jar
-rw-rw-r-- 1 jvs jvs   439876 Nov 28 14:25 sux4j-5.1.5.jar
-rw-rw-r-- 1 jvs jvs   379707 Nov 28 14:25 webgraph-3.6.7.jar
-rw-rw-r-- 1 jvs jvs   294580 Nov 28 14:25 webgraph-big-3.6.3.jar
```

For comparison, here's the 1.5.0 baseline.  If we can prune out any other spurious dependencies, that would be good too.  

And I'm thinking a separate jgrapht-webgraph module (dependent on jgrapht-opt) would be preferable to throwing everything directly into jgrapht-opt.

```
-rw-rw-r-- 1 jvs jvs   337869 Feb  5  2020 antlr4-runtime-4.8-1.jar
-rw-rw-r-- 1 jvs jvs   503880 Feb 23  2020 commons-lang3-3.9.jar
-rw-rw-r-- 1 jvs jvs   207030 Feb 23  2020 commons-text-1.8.jar
-rw-rw-r-- 1 jvs jvs 19011744 Feb 23  2020 fastutil-8.3.1.jar
-rw-rw-r-- 1 jvs jvs  2788302 Feb  5  2020 guava-28.2-jre.jar
-rw-rw-r-- 1 jvs jvs  1116840 Jun 15 15:38 jgrapht-core-1.5.0.jar
-rw-rw-r-- 1 jvs jvs    36037 Jun 15 15:40 jgrapht-demo-1.5.0.jar
-rw-rw-r-- 1 jvs jvs     5610 Jun 15 15:40 jgrapht-ext-1.5.0.jar
-rw-rw-r-- 1 jvs jvs    36574 Jun 15 15:40 jgrapht-guava-1.5.0.jar
-rw-rw-r-- 1 jvs jvs   354660 Jun 15 15:40 jgrapht-io-1.5.0.jar
-rw-rw-r-- 1 jvs jvs    28576 Jun 15 15:40 jgrapht-opt-1.5.0.jar
-rw-rw-r-- 1 jvs jvs   799664 Feb 23  2020 jgraphx-4.0.5.jar
-rw-rw-r-- 1 jvs jvs   152480 Feb 23  2020 jheaps-0.13.jar
```

**Comment by jsichi** [comment]
@vigna can you update the WebGraph home page with the news about the dual licensing?  I see it in the changelog but not on the main page.

**Comment by vigna** [comment]
> @vigna can you update the WebGraph home page with the news about the dual licensing? I see it in the changelog but not on the main page.

Thank you for noting this. I fixed the site, but the overview text will need a further release.

**Comment by vigna** [comment]
> Below is a list of the jars produced by `mvn package` off of the webgraph branch. We need exclusions for the jung stuff at a minimum.

Well, it turns out in the last 15 years we forgot to clean up our dependencies. A lot of those are actually not necessary. Let me complete the process...

**Comment by d-michail** [comment]
@jsichi I agree, having a separate `jgrapht-webgraph` seems like a good idea.  Sebastiano also has some succinct graph implementations using sux4j. How would these fit into the picture?

**Comment by vigna** [comment]
OK, these are the new dependencies:

```
jheaps-0.13.jar
jgrapht-io-1.5.1-SNAPSHOT.jar
antlr4-runtime-4.8-1.jar
commons-text-1.8.jar
commons-lang3-3.9.jar
jgrapht-ext-1.5.1-SNAPSHOT.jar
jgraphx-4.1.0.jar
jgrapht-guava-1.5.1-SNAPSHOT.jar
guava-29.0-jre.jar
jgrapht-demo-1.5.1-SNAPSHOT.jar
jgrapht-opt-1.5.1-SNAPSHOT.jar
fastutil-8.4.3.jar
dsiutils-2.6.15.jar
jsap-2.1.jar
slf4j-api-1.7.30.jar
logback-core-1.2.3.jar
logback-classic-1.2.3.jar
commons-configuration2-2.7.jar
commons-logging-1.2.jar
commons-math3-3.6.1.jar
sux4j-5.2.1.jar
commons-collections4-4.4.jar
webgraph-3.6.8.jar
webgraph-big-3.6.4.jar
```

Of these, the new ones with respect to the baseline are

```
dsiutils-2.6.15.jar
jsap-2.1.jar
slf4j-api-1.7.30.jar
logback-core-1.2.3.jar
logback-classic-1.2.3.jar
commons-configuration2-2.7.jar
commons-logging-1.2.jar
commons-math3-3.6.1.jar
sux4j-5.2.1.jar
commons-collections4-4.4.jar
webgraph-3.6.8.jar
webgraph-big-3.6.4.jar
```

These are all essential, as all our software is based on fastutil/dsiutils/Sux4J, we use JSAP for line parsing, Collections for predicates, Logback for logging and Math3 for some distribution and PRNG interfaces.

I have also updated all Commons to their newest versions in our software, so in case you start using some of them you won't get duplicate libraries (e.g., lang and lang3 or collections and collections4)

If you prefer a separate project, I'd go for a jgrapht-dsi where I would put stuff based on WebGraph and Sux4J.

**Comment by jsichi** [comment]
Can we go with jgrapht-dsi-webgraph for the module name to make it a little more specific and salient, while still encompassing?

A few more licensing headache questions:

* The DSI utilities page currently says LGPL, as does sux4j.  But the actual license appears to be GPLv3 + GCC Runtime Exception.  I think the exception means this should be compatible with redistribution under either our LGPLv2.1 or EPL 2.0; just want to confirm.
* The JSAP license says LGPL v2.1.  This won't work for redistribution under EPL 2.0.  Is JSAP actually necessary for use at runtime purely as a library?  I am guessing command line argument parsing is only needed for utilities.  If that's true, then let's add an exclusion to keep things simple.

**Comment by vigna** [comment]
Yes, it's OK for me. I'm rewriting the classes—I'm opting for a different, single design based on fastutil new type-specific pairs.

Mmmmhhh... they are LGPLv3+, and the manual says

"If you create a program that links through a shared library mechanism to a work that is separately distributed under LGPLv3, then you can distribute the resultant program under a license of your choice and you need not convey the LGPLv3’d work’s source code. If you distribute the library along with your program, or are the separate distributor of the work in another context or as another product, you must distribute its corresponding source under the terms of LGPLv3 or GPLv3-or-later."

So that should be OK AFAIK. JSAP can be eliminated—it's just for parsing the command line options. It's not used anywhere else.

BTW, thank you for checking all this—I'm fixing stuff that needed to be fixed since decades :).

**Comment by vigna** [comment]
Brief update: additional dependencies have been reduced a lot:

dsiutils-2.6.15.jar
jsap-2.1.jar
slf4j-api-1.7.30.jar
logback-core-1.2.3.jar
logback-classic-1.2.3.jar
commons-math3-3.6.1.jar
sux4j-5.2.1.jar
webgraph-3.6.8.jar
webgraph-big-3.6.4.jar

The problem is that JSAP cannot be removed—the class loader won't load classes containing references to it. I'm contacting the author, which has released other software under the ASF, to see whether he would release JSAP under the ASF or another linkable license like LGPL3+. The last release is from 2013, so I don't think he'll object to that.

**Comment by vigna** [comment]
BTW, why LGPL2.1+ won't work with the EPL? You are not distributing the binaries of JSAP anyway.

**Comment by jsichi** [comment]
I try to think of it from the point of view of a downstream open source project (or commercial product).  Assuming they haven't made any changes to JGraphT (nor to its dependencies), we want them to be able to produce complete self-contained redistributable binary releases (with or without source code).

**Comment by vigna** [comment]
OK, then, yes, we have a problem :).

**Comment by vigna** [comment]
OK, now I see why you were mentioning the runtime exception. Someone asked me to put Sux (the C++ version of my succinct libraries) under the GPLv3+ plus the runtime exception, which apparently makes everybody happy. But that was never passed onto Sux4J. Since this makes everybody happy, I'll add to Sux4J/DSI Utilities the runtime exception.

**Comment by vigna** [comment]
OK, the JSAP author is happy to re-release everything under a dual LGPL2.1+ + ASF 2.0 license.

**Comment by jsichi** [comment]
That's good news.  Periodically, we get questions from downstream developers about licensing, so whatever we can do to keep the answer as simple as possible is very helpful.

**Comment by vigna** [comment]
I'll relicense also the DSI Utilities and Sux4J as LGPL2.1+ / ASF 2.0 so also those are not a problem. The GNU Runtime Exception is only valid of C/C++ code compiled with GNU C. In the end there will be much fewer dependencies and all licenses will allow embedding.

**Comment by d-michail** [comment]
Great! 

@jsichi  Should we go for one package `jgrapht-dsi` which contains both webgraph and sux4j implementations? Apart from performance optimizations, the sux4j adapters are probably also ready.

**Comment by jsichi** [comment]
Yes, but as mentioned previously, I think we should call it jgrapht-dsi-webgraph since dsi by itself may not be obvious.

**Comment by vigna** [comment]
I see. Do you prefer then a jgrapht-dsi-webgraph and a jgrapht-dsi-sux4j ? The two implementations are completely independent  (the second implementation is not an adapter). Maybe jgrapht-comp, to allude at compression? The advantage of jgraph-dsi was that it would host different techniques developed here.

**Comment by jsichi** [comment]
If we want to keep everything in one module, then how about jgrapht-unimi-dsi?  Googling for "unimi dsi" takes you to the right place, but just "dsi" currently takes you to a Nintendo product :)

**Comment by vigna** [comment]
😂. Note that theoretically our department now is DI, not DSI, but all the software was developed for the DSI and has package it.unimi.dsi, so I think jgrapht-unimi-dsi is perfectly fine. I also modified the licensing of the DSI Utilities and Sux4J similarly to WebGraph (LPGL2+1 + ASF 2.0) so that should not be a problem anymore. Waiting for Marty to distribute the binaries for JSAP under the new license. If you're OK with merging in the mean time, let's go. Otherwise, I'll pester him a little bit.

**Comment by d-michail** [comment]
I think one module is better than two. If the sux4j stuff are ready, maybe you can merge them in the same branch? Then, I will try to find some time to create (copy-paste) the new module.

**Comment by d-michail** [comment]
I have a first version of the new module. I ended up using `org.jgrapht.webgraph` for the actual java package. I am not sure `org.jgrapht.unimi.dsi.webgraph` looks good. If you think so, or have a better idea, I can change it.

@vigna If you make any additional changes or finish the sux4j, feel free to merge the changes in this branch.

**Comment by vigna** [comment]
Seems perfect. Why don't we just merge in this one, which is pretty mature, and you leave me playing a bit still with the Sux4J implementation?

**Comment by d-michail** [comment]
Ok, I will try to also update the README with the new dependencies, etc and afterwards merge.

**Comment by vigna** [comment]
I'm waiting for this to be merged before adding the Sux4J stuff—in this way the module etc. will be right and already there.

**Comment by d-michail** [comment]
I am adding the latest changes in the README and updating license files. We will have to postpone merging until a compatible jsap jar is uploaded in maven central.

Also webgraph seems to be GPL3 in maven central. 

Would it also be possible to completely remove the loggers (not the api) but the implementations?

**Comment by vigna** [comment]
I'll try to pester Marty, but he did not answer to my last email. I guess it is onto other things, and also Christmas is coming. 

The two logback jars are runtime only (they are not needed for compilation). I don't think you can eliminate them or SLF4J will die, but I have to check.

**Comment by vigna** [comment]
No, actually you can exclude the logback dependencies—the user will just get a warning and no logging.

I'm uploading updated WebGraph versions on Maven with a fixed POM with the correct licensing information.

**Comment by d-michail** [comment]
Another possibility would be to switch command line parsing library, maybe jcommander (https://jcommander.org/) which has an Apache license.

**Comment by vigna** [comment]
No,  we cannot do that. We use JSAP pervasively, and we like it a lot. It's by far the best solution for command-line parsing we've seen.

**Comment by vigna** [comment]
But,  Marty tells me it'll be soon.

**Comment by d-michail** [comment]
ok, no problem. I have not used it myself.

**Comment by vigna** [comment]
OK, we have a solution: at the following Maven coordinates:
```
it.unimi.di:jsap-20210129.jar
```
there is a snapshot of the current Github repository for JSAP (https://github.com/martylamb/jsap). The repository is ASF 2.0-licensed—it's just that Marty has some issues in publishing the jar under the com.martiansoftware group id. Since the software is fixed from 2014 it does not really matter. By using the coordinates above in the POM as a dependency before our stuff, Maven will use the jar above and everybody will be happy. I'm re-releasing slowly our stuff with the same jar for convenience (and maybe in the future with an official JSAP jar), but there's no need to wait.

**Comment by d-michail** [comment]
@vigna  I was playing around with this implementation. I think there is a small bug when calling `.copy()` but the graph does not contain the transpose. We need to add a null check there otherwise you get a `NullPointerException`.

**Comment by vigna** [comment]
You're totally right, I don't have a unit test invoking copy() on a forward-only instance... it should be

```
transpose != null ? transpose.copy() : null
```

**Comment by vigna** [comment]
I also completed the release of our whole stack without the LGPL 2.1, so to be super sure that no spurious dependency on the LGPL 2.1 JSAP appears you can update fastutil to 8.5.1, dsiutils to 2.6.17, Sux4J to 5.2.3, webgraph to 3.6.10 and webgraph-big to 3.6.5.

**Comment by vigna** [comment]
OK, this is mergeable for me.

**Comment by d-michail** [comment]
Yep, for me too. @jsichi could you merge if everything looks fine?

File: jgrapht-unimi-dsi/src/test/java/org/jgrapht/webgraph/ImmutableDirectedGraphAdapterTest.java [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+        assertFalse(v.contains(3));
+        assertEquals(new IntOpenHashSet(v.iterator()), new IntOpenHashSet(new int[] { 0, 1, 2 }));
+        assertEquals(1, a.getEdgeSource(IntIntPair.of(1, 2)).longValue());
+
+        assertEquals(2, a.getEdgeTarget(IntIntPair.of(1, 2)).longValue());
+
+        assertEquals(1.0, a.getEdgeWeight(IntIntPair.of(1, 2)), 0);
+        a.setEdgeWeight(IntIntPair.of(0, 1), 1);
+
+        /*
```
Comment (jsichi) [inline jgrapht-unimi-dsi/src/test/java/org/jgrapht/webgraph/ImmutableDirectedGraphAdapterTest.java:230]: Remove commented-out code

File: jgrapht-unimi-dsi/src/test/java/org/jgrapht/webgraph/ImmutableDirectedGraphAdapterTest.java [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+        m.addArc(1, 2);
+        m.addArc(2, 2);
+
+        final ImmutableGraph g = m.immutableView();
+        final ImmutableGraph t = Transform.transpose(g);
+        final File basename = storeTempGraph(g);
+        final EFGraph ef = EFGraph.load(basename.toString());
+
+        /*
+         * Assertions.assertThrows(IllegalArgumentException.class, () -> {
```
Comment (jsichi) [inline jgrapht-unimi-dsi/src/test/java/org/jgrapht/webgraph/ImmutableDirectedGraphAdapterTest.java:112]: Remove commented-out code

File: jgrapht-unimi-dsi/src/test/java/org/jgrapht/webgraph/ImmutableDirectedBigGraphAdapterTest.java [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+        assertEquals(
+            new LongOpenHashSet(v.iterator()), new LongOpenHashSet(new long[] { 0L, 1L, 2L }));
+
+        assertEquals(1, a.getEdgeSource(LongLongPair.of(1L, 2L)).longValue());
+        assertEquals(2, a.getEdgeTarget(LongLongPair.of(1L, 2L)).longValue());
+
+        assertEquals(1, a.getEdgeWeight(LongLongPair.of(1L, 2L)), 0);
+        a.setEdgeWeight(LongLongPair.of(0L, 1L), 1);
+
+        /*
```
Comment (jsichi) [inline jgrapht-unimi-dsi/src/test/java/org/jgrapht/webgraph/ImmutableDirectedBigGraphAdapterTest.java:233]: Remove commented-out code

File: jgrapht-unimi-dsi/src/test/java/org/jgrapht/webgraph/ImmutableDirectedBigGraphAdapterTest.java [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+        m.addArc(1, 2);
+        m.addArc(2, 2);
+
+        final it.unimi.dsi.webgraph.ImmutableGraph mg = m.immutableView();
+        final ImmutableGraph g = ImmutableGraph.wrap(mg);
+        final ImmutableGraph t = ImmutableGraph.wrap(Transform.transpose(mg));
+        final File basename = storeTempGraph(g);
+        final EFGraph ef = EFGraph.load(basename.toString());
+
+        /*
```
Comment (jsichi) [inline jgrapht-unimi-dsi/src/test/java/org/jgrapht/webgraph/ImmutableDirectedBigGraphAdapterTest.java:113]: Remove commented-out code

**Review by jsichi — COMMENTED** [review]
Allllmost there...I did a mvn clean package and ended up with two versions of jsap in the jgrapht-dist:

-rw-rw-r-- 1 jvs jvs    90009 Feb  2 18:00 jsap-20210129.jar
-rw-rw-r-- 1 jvs jvs    69344 Nov 28 14:25 jsap-2.1.jar

I'm guessing this is due to what Seba mentioned about some of the dependency updates still being in progress, so we can use another exclude to drop the jsap-2.1.jar, right?

**Comment by vigna** [comment]
OK, that shouldn't really happen. Could you please do a dependency analysis with Maven and check who's pulling in JSAP 2.1? That shouldn't really happen. In any case, yes, you can exclude it.

**Comment by vigna** [comment]
Got it. webgraph-big should be 3.6.6. I made a typo when I suggested you the new versions.

fastutil is now at 8.5.2 (8.5.0 was a release with a ton of new code, so we're flushing out bugs quickly).

**Comment by jsichi** [comment]
Thanks, I'll merge once Travis CI is resolved.

**Comment by jsichi** [comment]
Hurray!  Github actions are ready now so I'll merge.

**Comment by jkinable** [comment]
@vigna @d-michail 
Could you include a demo class in our [wiki](https://github.com/jgrapht/jgrapht/wiki) to demonstrate the usage of the adapter? The demo class should also further explain the purpose/benefit of the webgraph adapter. Demo some cool application(s). In particular: what can jgrapht users do with the addition of webgraph that they couldn't do before, and, also the other way around, what new things can webgraph users do now they have access to jgrapht routines?

Other than a demo, what are 'standard' tasks/queries/computations that webgraph users perform on webgraphs? Does JGraphT support those algorithms, and do our algorithms scale sufficiently to webgraph sizes?

Finally, it might be useful to reference JGraphT on https://webgraph.di.unimi.it/.

I could create an issue to track progress on the above? It's important that we add context and some examples, if we want our users to adopt and use the new adapter classes.

**Comment by vigna** [comment]
Yes, I can do that. Presently JGraphT users are limited to graphs that fit the main memory, and for web graphs the WebGraph format uses 10-20 times less space than a trivial representation, and probably one-two orders of magnitude less than a JGraphT default representation. So if users want to run on a laptop an algorithm on a graph with a billion nodes they can do it with the adapters (e.g., http://law.di.unimi.it/webdata/eu-2015/) but in no other way (the graph has >90B arcs).

There are interesting algorithms in JGraphT we have not implemented in WebGraph. I do not know whether they would scale.

Yep we were planning to refer to JGraphT but I didn't know about the 1.5.1 release up to 2 days ago :).

Certainly you can create the issue—it's great to keep track.

**Comment by vigna** [comment]
In the meantime, there's a blurb pointing to JGraphT [here](http://law.di.unimi.it/datasets.php) and all our "small" datasets have a serialized succinct downloadable JGraphT version.

**Pull request merged**